File: desktop-client/src/app/game/services/review-manager/review-manager.service.ts
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { PlayersDrawingsPopUpComponent } from '@app/game/components/players-drawings-pop-up/players-drawings-pop-up.component';
import { ReviewPopUpComponent } from '@app/game/components/review-pop-up/review-pop-up.component';
import { ValidityPopUpComponent } from '@app/game/components/validity-pop-up/validity-pop-up.component';
import { GameStateService } from '@app/game/services/game-state/game-state.service';
import { WebsocketService } from '@app/shared/services/websocket/websocket.service';
import { PlayerAnswerForReview } from '@common/interfaces/player-answer-for-review';
import { PlayerDrawingAnswer } from '@common/interfaces/player-drawing-answer';
import { QuestionAnswer } from '@common/interfaces/question-answer';
import { GameEvent } from '@common/socket-events/game-event';

@Injectable({
    providedIn: 'root',
})
export class ReviewManagerService {
    constructor(
        private readonly gameStateService: GameStateService,
        private readonly webSocketService: WebsocketService,
        private readonly dialog: MatDialog,
    ) {}

    setupManager() {
        this.setupQuestionAnswersEvent();
        this.setupPlayerReviewAnswerEvent();
        this.setupDialogCloseNextQuestion();
        this.setupDialogCloseResultsView();
        this.setupDrawingsEvent();
    }

    resetReviewManager() {
        this.dialog.closeAll();
        this.webSocketService.removeAllListeners(GameEvent.SEND_ANSWERS);
        this.webSocketService.removeAllListeners(GameEvent.SEND_REVIEWS);
        this.webSocketService.removeAllListeners(GameEvent.NEXT_QUESTION);
        this.webSocketService.removeAllListeners(GameEvent.GAME_FINISHED);
        this.webSocketService.removeAllListeners(GameEvent.GET_DRAWINGS);
    }

    private setupQuestionAnswersEvent() {
        this.webSocketService.on(GameEvent.SEND_ANSWERS, (questionAnswer?: QuestionAnswer) => {
            if (!questionAnswer) return;
            this.gameStateService.playerScore = questionAnswer.score;
            this.showAnswerPlayer(questionAnswer);
        });
    }

    private setupDrawingsEvent() {
        this.webSocketService.on(GameEvent.GET_DRAWINGS, (playersDrawings?: PlayerDrawingAnswer[]) => {
            if (!playersDrawings) return;
            this.showDrawings(playersDrawings);
        });
    }

    private setupPlayerReviewAnswerEvent() {
        this.webSocketService.on(GameEvent.SEND_REVIEWS, (playerAnswers?: PlayerAnswerForReview[]) => {
            if (playerAnswers) {
                this.showReviewOpenEndedAnswersPopUp(playerAnswers);
            }
        });
    }

    private setupDialogCloseNextQuestion() {
        this.webSocketService.on(GameEvent.NEXT_QUESTION, () => {
            this.dialog.closeAll();
        });
    }

    private setupDialogCloseResultsView() {
        this.webSocketService.once(GameEvent.GAME_FINISHED, () => {
            this.dialog.closeAll();
        });
    }

    private showAnswerPlayer(correctAnswers: QuestionAnswer) {
        const dialogRef = this.dialog.open(ValidityPopUpComponent, {
            disableClose: true,
        });
        dialogRef.componentInstance.setupPopUp(correctAnswers, this.gameStateService.isEliminated);
    }

    private showDrawings(playersDrawings: PlayerDrawingAnswer[]) {
        const dialogRef = this.dialog.open(PlayersDrawingsPopUpComponent, {
            disableClose: true,
        });
        dialogRef.componentInstance.setupPopUp(playersDrawings);
    }

    private showReviewOpenEndedAnswersPopUp(playerAnswers: PlayerAnswerForReview[]) {
        const dialogRef = this.dialog.open(ReviewPopUpComponent, {
            disableClose: true,
        });
        dialogRef.componentInstance.setPopUp(playerAnswers);
    }
}

File: mobile_client/lib/services/game/review_manager_service.dart
import 'package:flutter/material.dart';
import 'package:mobile_client/constants/enum/page_url.dart';
import 'package:mobile_client/constants/socket_events/game_event.dart';
import 'package:mobile_client/models/game/player_answer_for_review.dart';
import 'package:mobile_client/models/game/player_drawing_answer.dart';
import 'package:mobile_client/models/game/question_answer.dart';
import 'package:mobile_client/services/game/game_state_service.dart';
import 'package:mobile_client/services/websocket_service.dart';
import 'package:mobile_client/utils/app_navigator.dart';
import 'package:mobile_client/widgets/game/player_review_pop_up.dart';
import 'package:mobile_client/widgets/game/players_drawings_pop_up.dart';
import 'package:mobile_client/widgets/game/validity_pop_up/validity_pop_up.dart';

class ReviewManagerService {
  final GameStateService _gameStateService;
  final WebSocketService _webSocketService;
  BuildContext? _validityPopUp;

  ReviewManagerService(
    this._gameStateService,
    this._webSocketService,
  );

  void setupManager() {
    _setupQuestionAnswersEvent();
    _setupPlayerAnswerReviewEvent();
    _setupDialogCloseNextQuestion();
    _setupDialogCloseResultsView();
    _setupDrawingsEvent();
  }

  void resetReviewManager() {
    _webSocketService.removeAllListeners(GameEvent.sendAnswers.name);
    _webSocketService.removeAllListeners(GameEvent.sendReviews.name);
    _webSocketService.removeAllListeners(GameEvent.nextQuestion.name);
    _webSocketService.removeAllListeners(GameEvent.gameFinished.name);
    _webSocketService.removeAllListeners(GameEvent.getDrawings.name);
  }

  void _setupQuestionAnswersEvent() {
    _webSocketService.on<Map<String, dynamic>?>(GameEvent.sendAnswers.name,
        (json) {
      if (json == null) return;
      final questionAnswer = QuestionAnswer.fromJson(json);

      _gameStateService.playerScore.value = questionAnswer.score;
      _showAnswerPlayer(questionAnswer);
    });
  }

  void _setupDrawingsEvent() {
    _webSocketService.on<List<dynamic>?>(GameEvent.getDrawings.name, (json) {
      if (json == null) return;
      final playersDrawings = json
          .map((playerDrawing) => PlayerDrawingAnswer.fromJson(playerDrawing))
          .toList();
      _showDrawings(playersDrawings);
    });
  }

  void _setupPlayerAnswerReviewEvent() {
    _webSocketService.on<List<dynamic>>(GameEvent.sendReviews.name, (json) {
      final playerAnswers = json
          .map((playerAnswer) => PlayerAnswerForReview.fromJson(playerAnswer))
          .toList();
      _showReviewAnswersPopUp(playerAnswers);
    });
  }

  void _setupDialogCloseNextQuestion() {
    _webSocketService.on(GameEvent.nextQuestion.name, (_) {
      _clearPopUp();
    });
  }

  void _setupDialogCloseResultsView() {
    _webSocketService.once(GameEvent.gameFinished.name, (_) {
      _clearPopUp();
      AppNavigator.go('/${PageUrl.results.name}');
    });
  }

  void _showAnswerPlayer(QuestionAnswer questionAnswer) {
    if (AppNavigator.navigatorKey.currentContext == null) return;
    showDialog(
        context: AppNavigator.navigatorKey.currentContext!,
        barrierDismissible: false,
        builder: (context) {
          _validityPopUp = context;
          return ValidityPopUp(
              questionAnswer: questionAnswer,
              isEliminated: _gameStateService.isEliminated.value);
        });
  }

  void _clearPopUp() {
    if (_validityPopUp == null) return;
    AppNavigator.pop(AppNavigator.navigatorKey.currentContext!);
    _validityPopUp = null;
  }

  void _showDrawings(List<PlayerDrawingAnswer> playersDrawings) {
    if (AppNavigator.navigatorKey.currentContext == null) return;
    showDialog(
        context: AppNavigator.navigatorKey.currentContext!,
        barrierDismissible: false,
        builder: (context) {
          _validityPopUp = context;
          return PlayersDrawingsPopUp(playersDrawings: playersDrawings);
        });
  }

  void _showReviewAnswersPopUp(List<PlayerAnswerForReview> playerAnswers) {
    if (AppNavigator.navigatorKey.currentContext == null) return;
    showDialog(
        context: AppNavigator.navigatorKey.currentContext!,
        barrierDismissible: false,
        builder: (context) {
          _validityPopUp = context;
          return PlayerReviewPopUp(answers: playerAnswers);
        });
  }
}
